src/core/utils/clients/app-manager.ts

```diff
@@ -0,0 +1,409 @@
+      extraProgramPages: Number(app.params.extraProgramPages ?? 0),
+      globalState: AppManager.decodeAppState(app.params.globalState ?? []),
+    }
+  }
+
+  async getGlobalState(appId: bigint): Promise<Record<string, AppState>> {
+    const appInfo = await this.getById(appId)
+    return appInfo.globalState
+  }
+
+  async getLocalState(appId: bigint, address: string): Promise<Record<string, AppState>> {
+    const appInfo = await this.algodClient.accountApplicationInformation(address, appId)
+
+    if (!appInfo.appLocalState?.keyValue) {
+      throw new Error("Couldn't find local state")
+    }
+
+    return AppManager.decodeAppState(appInfo.appLocalState.keyValue)
+  }
+
+  async getBoxNames(appId: bigint): Promise<BoxName[]> {
+    const boxResult = await this.algodClient.getApplicationBoxes(appId)
+    return boxResult.boxes.map((b) => {
+      return {
+        nameRaw: new Uint8Array(Buffer.from(b.name)),
+        nameBase64: b.name,
+        name: Buffer.from(b.name).toString('utf-8'),
+      }
+    })
+  }
+
+  async getBoxValue(appId: bigint, boxName: Uint8Array): Promise<Uint8Array> {
+    // Algod expects goal-arg style encoding for box name query param in 'encoding:value'.
+    // However our HTTP client decodes base64 automatically into bytes for the Box model fields.
+    // The API still requires 'b64:<base64>' for the query parameter value.
+    const processedBoxName = `b64:${Buffer.from(boxName).toString('base64')}`
+
+    const boxResult = await this.algodClient.getApplicationBoxByName(appId, {
+      name: processedBoxName,
+    })
+    return new Uint8Array(Buffer.from(boxResult.value))
+  }
+
+  async getBoxValues(appId: bigint, boxNames: Uint8Array[]): Promise<Uint8Array[]> {
+    const values: Uint8Array[] = []
+    for (const boxName of boxNames) {
+      values.push(await this.getBoxValue(appId, boxName))
+    }
+    return values
+  }
+
+  private static ensureDecodedBytes(bytes: Uint8Array): Uint8Array {
+    try {
+      const str = Buffer.from(bytes).toString('utf8')
+      if (
+        str.length > 0 &&
+        /^[A-Za-z0-9+/]*={0,2}$/.test(str) &&
+        (str.includes('=') || str.includes('+') || str.includes('/') || (str.length % 4 === 0 && str.length >= 8))
+      ) {
+        const decoded = Buffer.from(str, 'base64')
+        if (!decoded.equals(Buffer.from(bytes))) {
+          return new Uint8Array(decoded)
+        }
+      }
+    } catch {
+      // Not valid UTF-8 or base64, return as-is
+    }
+    return bytes
+  }
+
+  static decodeAppState(state: TealKeyValueStore): Record<string, AppState> {
+    const stateValues: Record<string, AppState> = {}
+
+    for (const stateVal of state) {
+      const keyRaw = new Uint8Array(Buffer.from(stateVal.key, 'base64'))
+      const keyBase64 = stateVal.key
+      const keyString = Buffer.from(keyRaw).toString('base64')
+
+      // TODO: we will need to update the algod client to return int here
```

Comment:
I'm assuming you've raised all these TODOs up to Al as feeback on the http client PR?